src/core_plugins/kibana/public/dashboard/dashboard.js

```diff
@@ -226,6 +240,43 @@ app.directive('dashboardApp', function ($injector) {
         $scope.indexPatterns = dashboardState.getPanelIndexPatterns();
       };
 
+      $scope.filter = function (field, value, operator, index) {
+        if ($scope.model.query.language === 'lucene') {
+          filterManager.add(field, value, operator, index);
+        }
+
+        if ($scope.model.query.language === 'kuery') {
+          const fieldName = _.isObject(field) ? field.name : field;
+          const kueryAST = fromKueryExpression($scope.model.query.query);
+          const newAST = addNode(
+            kueryAST,
+            nodeTypes.match.buildNode({ field: fieldName, values: value, operation: operator })
+          );
+          $scope.updateQuery({ query: toKueryExpression(newAST), language: 'kuery' });
+        }
+      };
+
+
+      $scope.$watch('model.query', (newQuery) => {
+        $scope.model.query = migrateLegacyQuery(newQuery);
+        dashboardState.applyFilters($scope.model.query, filterBar.getFilters());
+        $scope.refresh();
+      });
+
+      $scope.$watch(() => dashboardState.getAppState().$newFilters, function (filters = []) {
```

Comment:
You'll see similar watches to this one in each app. I'd like to turn this logic into a more generic "query/filter manipulation" API that can be reused. I plan to tackle that in a separate PR because we'll also need to get feedback from the visualize team to see if it fits their needs in the visualize refactor.